[Swift] Swift implementation 🎉🎉

mustiikhalil opened the pull request:
Good afternoon, 

@mzaks and I hope with this PR we would be able to add an official implementation for swift, it's heavily inspired by the `C++` and `C#` implementations. There are some elements that are still missing such as the code generator and also documentation for the code base, however I found that we can implement those after getting an initial review for the swift code base, and if it lives up to the standards. 

The `FlatBuffer` class  uses apples underlaying `UnsafeMutableRawPointer` which allows is to write directly to the memory, and lefts a lot of heavy weight from our shoulders, since this implementation comes which the following function `storeBytes` with takes a value and it's Bytes representation and just adds it to the buffer. `UnsafeMutableRawPointer` the name is a bit misleading since it's actually safe to use if everything writing the bytes is well structured. and since we don't really allow the user to get to the underlying adding and pushing to the buffer and only allowing them to do so through the `builder` it would be completely safe to use it.

The `FlatBuffersBuilder` class uses the same underlying logic as `C++` and `C#`, but unlike `C#` it relies on generics to actually add bytes and elements into the buffer. we took a different approach regarding adding the structs into the buffer, since swift uses the `storeBytes` to actually store the struct into the buffer, and since `FlatBuffers` do not allow string in `Structs` we took that into our advantage. 

```swift
struct Vec2: Writeable {
    var _x: Float32
    var _y: Float32
    var _z: Float32
    var _c: Color2

     init(x: Float32, y: Float32, z: Float32, color: Color2) { _c = color; _x = x; _y = y; _z = z }

     static func createVec2(_ bb: FlatBuffersBuilder, v: Vec2) -> Offset<UOffset> {
        return bb.create(struct: v)
    }
}
```
the following struct will be able to be directly inserted into the buffer, and we will be implementing a reader struct that will read the Writable struct as shown below. This will have the same concept as the builder Objects in c++. The readers were implemented to follow the `C#` way of fetching elements from the buffer.

```swift
struct Vec2_Read: Readable {
    private var __p: Struct
    init(_ fb: FlatBuffer, o: Int32) { __p = Struct(bb: fb, position: o) }
    var c: Color2 { return Color2(rawValue: __p.readBuffer(of: Int32.self, at: 12)) ?? .red }
    var x: Float32 { return __p.readBuffer(of: Float32.self, at: 0)}
    var y: Float32 { return __p.readBuffer(of: Float32.self, at: 4)}
    var z: Float32 { return __p.readBuffer(of: Float32.self, at: 8)}
}
```

All the test case are passing, and they were created by the `flatc` code generator for both `C++` and `C#`, and the basic implementation of swift was verified against those two since the code generator isn't implemented yet. 

what's missing:
1- Code Generator
2- Documentation
3- Benchmarking

when merging this PR we would be closing the following issues, Closes #5504. and I will be rebasing the commit before merging for sure

dbaileychess commented:
Can you add some tests that read the `tests/monsterdata_test.mon` to ensure that data written by another language is correctly read by Swift?

mustiikhalil commented:
Definitely would do that, I just want to know if the `monsterdata_test.mon` data is mapped to the following `monster_test.fbs`? Since I need to know what are the actual objects with the buffer.

dbaileychess commented:
Yep, `monster_test.fbs` is the schema and `monsterdata_test.json` are the values that are used. You can also follow the other tests files for C#, Java, etc.. to see how they use the `monster_test.mon` file to test.

aardappel commented:
Thanks, this is exciting! Some quick thoughts from your readme, and then an initial code review will follow:

- Looking at the docs this is not clear, but I am guessing that `UnsafeMutableRawPointer` when it reads a 16/32/64-bit quantity, it will do so with machine-endianness ordering, which means it won't work correctly on big endian machines. This means you should either have explicit big-endian versions of most of these functions, or (and this is entirely acceptable) explicitly disable this code to run on big-endian.
- As for splitting a struct into 2 versions for reading and writing: this sounds acceptable to me, except I'd give the version for reading the shorter name, as it may be used more often. Or rather, reading code is more important to be "pretty" or native looking.
- There's no code generator? Meaning all current code was hard-coded for the test? That is useful for an initial look, but that makes it far off being a complete PR.

aardappel commented on swift/.swiftpm/xcode/xcshareddata/xcschemes/FlatBuffers.xcscheme:
```diff
@@ -0,0 +1,102 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Ideally we don't store IDE data in repo, not sure how unavoidable that is in the case of Swift. Does this code build from the command-line without XCode files?

mustiikhalil replied:
I believe we can remove that indeed, however we will need a way to make SPM work since it would be a nested file. since it would be looking for the Package.swift, so I think we might need to make it a git submodule

aardappel replied:
We currently don't have any submodules for this repo, so would be good to avoid starting that now :)

aardappel commented on tests/FlatBuffers.Test.Swift/Tests/FlatBuffers.Test.SwiftTests/FlatBuffersTests.swift:
```diff
@@ -0,0 +1,118 @@
+    }
+}
+
+class Country {
```
This seems a new test case specific to Swift? Important that once we have codegen, that it uses the existing tests.

mustiikhalil replied:
Yes, all the test cases were writing manually for swift since we don't have a codegen. however they were all written and compared with both `cpp` and `C#`

aardappel replied:
so? can these tests be removed now then? Or do you think they provide useful extra coverage?

mustiikhalil replied:
I think it's nice to have them for extra coverage.

aardappel commented on swift/Sources/FlatBuffers/Table.swift:
```diff
@@ -0,0 +1,60 @@
+        return o + _bb.read(def: Int32.self, position: Int(o)) + 4
+    }
+    
+    public func readBuffer<T: Scalar>(of type: T.Type, offset o: Int32) -> T {
```
I assume in Swift this `T.Type` argument is fully static and does not involve any runtime value?

mustiikhalil replied:
> A metatype type refers to the type of any type, including class types, structure types, enumeration types, and protocol types. https://docs.swift.org/swift-book/ReferenceManual/Types.html
This would request the codegen to specify the type of the readable object before reading it. As in `__p.readBuffer(of: Double.self, offset: 10)` since Double confirms to the type `Scalar`, however I am not sure if it would be fully static or not.

mustiikhalil replied:
On a side note, this is how `Swift` and `Apple` have implemented most of their methods, and it can be seen all over swift. so I think it would be pretty safe to use it when using the code generator, since users would be advised against changing it.

mustiikhalil replied:
Just found a talk from apple that might actually help out `https://developer.apple.com/videos/play/wwdc2016/416/`

mustiikhalil commented on swift/Sources/FlatBuffers/FlatBuffersBuilder.swift:
```diff
@@ -0,0 +1,385 @@
+    /// - Parameter initialSize:
+    public init(initialSize: Int32 = 1024, serializeDefaults force: Bool = false) {
+        guard initialSize > 0 else { fatalError(FlatbufferError.sizeIsZeroOrLess.errorDescription ?? "") }
+        guard isLitteEndian else { fatalError(FlatbufferError.endianCheck.errorDescription ?? "") }
```
> * Looking at the docs this is not clear, but I am guessing that `UnsafeMutableRawPointer` when it reads a 16/32/64-bit quantity, it will do so with machine-endianness ordering, which means it won't work correctly on big endian machines. This means you should either have explicit big-endian versions of most of these functions, or (and this is entirely acceptable) explicitly disable this code to run on big-endian.

Currently, indeed the code is disabled from running on Big-endian machines. In hopes that we would be implementing it whenever we get to have an issue regarding it.

aardappel replied:
That's fine.

mustiikhalil commented:
> * As for splitting a struct into 2 versions for reading and writing: this sounds acceptable to me, except I'd give the version for reading the shorter name, as it may be used more often. Or rather, reading code is more important to be "pretty" or native looking.

The struct names can be modified, it won't be a problem, when we get to have the code generator. 

> * There's no code generator? Meaning all current code was hard-coded for the test? That is useful for an initial look, but that makes it far off being a complete PR.

Yes, all the test cases were hard coded for the current tests, we wanted to get your opinion regarding the current implementation before going any further than what we currently have.

mustiikhalil commented:
@aardappel, @mzaks  and I were implementing the test cases for the swift, that adhere to the test cases that the FlatBuffers repository has. However, we ran into an issue: as explained in following [issue](https://github.com/mustiikhalil/FlatBuffer/issues/26). the short explanation would be the current implementation fo the swift doesn't allow us to next objects into structs, or have different types which is solved easily using the following:
```
withUnsafeBytes(of: &vec) { (buffer)  in
    _memory.advanced(by: writerIndex - len).copyMemory(from: buffer.baseAddress!, byteCount: buffer.count)
}
```
however as explained in the issue this will lead into a none zero padding, would that be an issue? 
example:

```
struct Test { a: short; b: byte }
struct Vec3 (force_align: 8) {
  x:float;
  y:float;
  z:float;
  test1:double;
  test2:Color;
  test3:Test;
}
```
would result into the following Binary representation for cpp and c#: 
`[0, 0, 128, 63, 0, 0, 0, 64, 0, 0, 64, 64, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 8, 64, 1, 0, 5, 0, 6, 0, 0, 0]`

however since we are writing and reading from memory in swift for structs this might lead to something like this
`[0, 0, 128, 63, 0, 0, 0, 64, 0, 0, 64, 64, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 8, 64, 1, 0, 5, 0, 6, 0, 0, 0]`

aardappel commented:
@mustiikhalil you're saying Swift's padding can generate garbage data for padding?

While I don't believe there is currently any code that expects padding to be 0, that is what all other languages write out, so Swift generating non-zero padding could cause problems for any code expecting binary equivalence of some sort. I'd say writing it in a way that guarantees zeroes is pretty important.

Having garbage there also makes output non-deterministic, and it makes compression of serialized data worse.

mustiikhalil commented:
yeah it makes sense, it's just in the structs that we face this issue, we can find other ways to implement it and fix it

mustiikhalil commented:
@aardappel following up on yesterdays issue, I managed to implement the following fix for the structs, however I am not sure if you would approve on it

```swift
struct Vec3Writer: Writeable {
    
    var memory: UnsafeMutableRawPointer
    static var size = 32
    static var alignment = 8

    init(x: Float32, y: Float32, z: Float32, test1: Double, color: Color_1, test: Test_1) {
    /// creates a memory that would store the struct in itself so it would be passed on later 
        memory = UnsafeMutableRawPointer.allocate(byteCount: Vec3Builder.size, alignment: Vec3Builder.alignment)
        memory.initializeMemory(as: UInt8.self, repeating: 0, count: Vec3Builder.size)
        memory.storeBytes(of: x, toByteOffset: 0, as: Float32.self)
        memory.storeBytes(of: y, toByteOffset: 4, as: Float32.self)
        memory.storeBytes(of: z, toByteOffset: 8, as: Float32.self)
        memory.storeBytes(of: test1, toByteOffset: 16, as: Double.self)
        memory.storeBytes(of: color.rawValue, toByteOffset: 24, as: Int8.self)
        memory.storeBytes(of: 0, toByteOffset: 25, as: Int8.self)
        memory.storeBytes(of: test.a, toByteOffset: 26, as: Int16.self)
        memory.storeBytes(of: test.b, toByteOffset: 28, as: Int8.self)
    }
}
let v = Vec3Writer(x: 1, y: 2, z: 3, test1: 3, color: .green, test: Test_1(a: 5, b: 6))

/// this is the internal call for copying the memory of the current implemented struct

buffer.copyMemory(v.memory, count: size)
```

mzaks commented:
> @mustiikhalil you're saying Swift's padding can generate garbage data for padding?
> 
> While I don't believe there is currently any code that expects padding to be 0, that is what all other languages write out, so Swift generating non-zero padding could cause problems for any code expecting binary equivalence of some sort. I'd say writing it in a way that guarantees zeroes is pretty important.
> 
> Having garbage there also makes output non-deterministic, and it makes compression of serialized data worse.

I was also surprised to see non 0 padding, but it is how it is.
The alternative is, what @mustiikhalil described in his latest reply - storing data value by value. It has it's benefits actually, as it is easier to provide BigEndian support. We are not dependent on Swift struct layout, which theoretically might change and be different from the one in FlatBuffers. And we can support custom FlatBuffers attributes on struct definitions, such as `force_align`.
The downside is that the generated code will become more complex and the performance of struct serialisation and probably deserialisation will be slower as we are not just doing a single memory copy. 

Regarding deserialisation. If we are doing a value by value serialisation than it makes sense to have value by value deserialisation as otherwise we might hit problem (the byte array representing a struct being not compatible with Swift struct representation).

aardappel commented:
Yes, most languages can't guarantee that their memory layout is the same as FlatBuffers, so should typically be done one by one. In C++ we have so macro trickery to check at compile time that the layout is the same, and so far we haven't found a compiler for which that fails, but if we did, we'd need per-field serialization even in C++.

As for `Vec3Writer`, I don't understand why this has to be struct, why not simply a free-standing function that gets a `FlatBufferBuilder` argument along with all the rest?

mustiikhalil commented:
@aardappel we created it as a struct so that we would be able to insert structs as a vector, which would have the alignment of the memory and also the size of the memory stripe. since we don't want to have variables that are laying around, 
```swift
static var size = 32
static var alignment = 8
```
that's why we implemented it with a struct to contain the values within the data structure 

but other than the struct, does the implementation of the `memory` and sending it to the `FlatBufferBuilder`?

mustiikhalil commented:
@aardappel currently @mzaks and I have almost finished everything regarding the swift implementation. we've implemented all the required APIs and also convenience APIs that would allow the codegen to implement the functions accordingly and without worrying about the index and positions of stuff. 
The `Table` implementation has `directRead` with ignores the position of an object, and also a `read` that respects it. that's all to minimize the codeGen work. 

 @dbaileychess as you requested earlier when we opened the PR the implementation of the test cases that will be reading from different languages, that is done using the `.mon` file you referenced earlier. currently we can write/read the object `Monster`. With our current exposed and implemented `APIs`. (all of this was done referencing the CSharp code) 

what's missing:
- Documentation
- Some functions in the table/struct implementation, including the following function implementation `TestarrayoftablesByKey` 
- codegen

aardappel commented:
@mustiikhalil does this have codegen now?

mustiikhalil commented:
@aardappel i just want a final review for the current swift APIs. so we start building the code gen.

aardappel commented:
@mustiikhalil reviewing a PR this size is a lot of work, so I'd prefer doing that again once you feel you've done everything you can.

Besides, API changes are probably even easier once you have code generation, and it is easier to see for me how it compares to the other languages when it is generating code for the exact same schemas and the same tests.

mustiikhalil commented:
@aardappel and @mzaks I have a couple of questions regarding the code gen and would love to get your input regarding the matter. I've built like 90% of the code generator, however I am not sure how to handle the namespaces? we don't want people to keep typing `MyGame.Example.Monster` each time they are gonna create a Monster object. and if we remove namespaces from the swift implementation, how do we call the new objects in swift? `Monster`, and `Monster1` 

what we are missing currently is the following in the code gen:
1- generating the comments in the fbs file
2- a create object function for the tables
3- a lookup function for the keys in the buffer

do we need to implement the function to create an object in swift? `CreateMonster(fbb: ...)`

mzaks commented:
> ... how to handle the namespaces? we don't want people to keep typing MyGame.Example.Monster each time they are gonna create a Monster object ...
Namespaces are a bit of a problem as Swift doesn't really have the concept. In Swift you either use Modules (but you can't have nested modules), or people also wrap definitions in empty enums, in order to mimic a name space. e.g.:

```
public enum MyGame {
  public enum Example {
    public struct Monster {
      ...
    }
  }
}
```

In order to avoid repeating your self with `MyGame.Example.Monster` you users can define type aliases:
```
typealias Monster = MyGame.Example.Monster
```

Which is not ideal, but in case that your schema makes use of namespacing and it is possible that Swift is not the only language using this schema, I am afraid this is the only way to make thing correct.

> do we need to implement the function to create an object in swift? CreateMonster(fbb: ...)
Not sure I understand the question.

mustiikhalil commented:
> > do we need to implement the function to create an object in swift? CreateMonster(fbb: ...)
> 
> Not sure I understand the question.

There are some tables that would have a constructor that will hold all the values in place example: 
```swift
static func createMonster(builder: inout FlatBuffersBuilder,
                                         position: Offset<UOffset>,
                                         hp: Int16,
                                         name: Offset<String>,
                                         inventory: Offset<UOffset>,
                                         color: Color3,
                                         weapons: Offset<UOffset>,
                                         equipment: Equipment = .none,
                                         equippedOffset: Offset<Weapon>,
                                         path: Offset<UOffset>) -> Offset<Monster> {
...
}
```
are we required to have this in the initial release?

aardappel commented:
Yes, use of namespaces should mimic the other languages, since we're not changing the schemas for Swift.

To what extend is this enum trick standard? Should we use it? Since this is generated code, some verbosity is ok, so simply using full names everywhere should be ok.

For the user, if there is no way to do "using namespace" in Swift, they are going to have to deal with the long names. We can't generate typealiases for them, since they may clash.

Having the 1 call `createMonster` is certainly nice to have, but not required.

mustiikhalil commented:
It's either we use Enums or structs for that and as @mzaks already suggested I think the enums will do the job, since apple does use it to nest classes in their `Combine Framework`. so I think it's doable. The typealiases will be made by the users, since as you mentioned it's kinda impossible to generate ones that will not clash.

mzaks commented:
@mustiikhalil sorry for being so unresponsive. Will try to go through the generated code this week.

mustiikhalil commented:
@aardappel and @mzaks I just finished the implementation for the code Gen, and I do apologize in advance if the cpp code isn't as you expect it to be, the code Gen generates the swift file by `--swift`, and just like other languages, the mutable will only be generated if we use `--gen-mutable`.  Since `Xcode` requires the user to drop the code to it's navigator is I just simply added a file called `monster_test_generated.swift` that hosts all the generated code. 
The code is already tested on a linux system that's running swift, and that's why we have the 
```swift
#if os(Linux)
import CoreFoundation
#else
import Foundation
#endif
```
however the test cases, that are running on this PR are still failing, I am not sure why though

aardappel commented:
@mustiikhalil to make the C++ code more conform, first step is to run `clang-format` on your generator (see also `src/clang-format-git.sh`), then compare your code with that of the other generators (in particular, the C++ one).

And yes, would be good to have the tests pass first :)

mustiikhalil commented:
> @mustiikhalil to make the C++ code more conform, first step is to run `clang-format` on your generator (see also `src/clang-format-git.sh`), then compare your code with that of the other generators (in particular, the C++ one).

in the test cases I always see this `/usr/include/c++/8/bits/stl_vector.h` in some of the errors messages, is it frowned upon to use std::vector in cpp?

> And yes, would be good to have the tests pass first :)

Definitely! we don't want the swift branch to hinder the flow of the development.

mustiikhalil commented:
@aardappel and @mzaks I think it's safe to say everything is running as expected! the test cases are passing. There are a lot of more things that we will be bringing to swift in the future like gRPC, and json encoding but that's after having the following approved.

mzaks commented:
Hi @mustiikhalil I briefly checked the generated code. There is still one main problem, the name space generation is not implemented properly it just adds `}` character, which also makes the code not compilable.
I noticed another small thing that the table structs have `Start...` `Add...` `End...` API but do not have a convenience API to create a table in one call with all the values. However I am not sure if we decided to skip this API at all in previous discussions.

All in all, really great work @mustiikhalil!

mustiikhalil commented:
@mzaks I will definitely take a look at that. I will revisit that and try to fix the issue there, do you have a specific test case that breaks it? 

> Having the 1 call createMonster is certainly nice to have, but not required.
as @aardappel said earlier it's a nice to have, maybe in the future when we add the JSON encoding we can add that too

mzaks commented:
I order to give more insight in what I mean by namespace is not implemented correctly. When I generate with following fbs:
```
table Person1 {
	name: string;
}

root_type Person1;
```

I get following code:
```
// automatically generated by the FlatBuffers compiler, do not modify

import FlatBuffers

}

struct Person1: FlatBufferObject {
	private var _accessor: Table
	static func getRootAsPerson1(bb: FlatBuffer) -> Person1 { return Person1(Table(bb: bb, position: Int32(bb.read(def: UOffset.self, position: bb.reader)) + Int32(bb.reader))) }

	init(_ t: Table) { _accessor = t }
	init(_ bb: FlatBuffer, o: Int32) { _accessor = Table(bb: bb, position: o) }

	public var name: String? { let o = _accessor.offset(4); return o == 0 ? nil : _accessor.string(at: o) }
	public var nameSegmentArray: [UInt8]? { return _accessor.getVector(at: 4) }
	static func startPerson1(_ fbb: FlatBuffersBuilder) -> UOffset { fbb.startTable(with: 1) }
	static func add(name: Offset<String>, _ fbb: FlatBuffersBuilder) { fbb.add(offset: name, at: 0)  }
	static func endPerson1(_ fbb: FlatBuffersBuilder, start: UOffset) -> Offset<UOffset> { let end = Offset<UOffset>(offset: fbb.endTable(at: start)); return end }
}
```

You can see the `}` character after `import FlatBuffers` statement. It is as far as I understand generated by not fully implemented namespace generation code:
https://github.com/mustiikhalil/flatbuffers/blob/5da59119c08886da87eb237ed646d6d2515d8251/src/idl_gen_swift.cpp#L632

mzaks commented:
In the long run, it would be great if we would have a sample for Swift similar to other languages in the samples folder:
https://github.com/mustiikhalil/flatbuffers/tree/swift-implementation/samples

mustiikhalil commented:
I can add that later today actually :D I don't think it would take too much time to add a script to that file so it generates the swift implementation

mustiikhalil commented:
@mzaks I think I fixed the issue with the namespaces, I believe it was because I didn't use `*shouldGenerateExtension` and also now we save the depth we are in so we can close the namespaces properly

mustiikhalil commented:
@aardappel I've cleaned up the project, finished implementing the code gen and also addressed the issues @mzaks brought up. I think we are ready for a code review

aardappel commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,717 @@
+    for (auto it = struct_def.fields.vec.begin();
+         it != struct_def.fields.vec.end(); ++it) {
+      auto &field = **it;
+      std::string name = nameprefix + Name(field);
```
use auto where-ever possible.

mustiikhalil replied:
the `nameprefix` is actually a string that's being passed from a different function, when I tried to make it auto that didn't work as you wanted it to be

aardappel commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,717 @@
+               offsetReader + "))";
+    }
+
+    code_ += doubleSpacing + "if comp > 0 {";
```
use `snake_case` for variables, here and elsewhere

aardappel commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,717 @@
+           optional + " { ";
+  }
+
+  std::string GenIndirect(std::string reading) {
```
Please inline all functions called just once.

mustiikhalil replied:
Do you mean remove them outside the class? and keep them in the swift namespace?

aardappel commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,717 @@
+  std::string GenTypeBasic(const Type &type, bool CanOverride) const {
+    // clang-format off
+        static const char * const swift_type[] = {
+#define FLATBUFFERS_TD(ENUM, IDLTYPE, \
```
indent pre-processor statements. Surround with `// clang-format: off` and `on` if this was caused by `clang-format`.

aardappel commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,717 @@
+            FLATBUFFERS_GEN_TYPES(FLATBUFFERS_TD)
+#undef FLATBUFFERS_TD
+        };
+        if (CanOverride) {
```
`snake_case` for variables.

aardappel commented on tests/FlatBuffers.Test.Swift/.gitignore:
```diff
@@ -0,0 +1,7 @@
+/.build
+/Packages
+/*.xcodeproj
+xcuserdata/
```
How does this relate to the same dirs ignored in the root?
Why is this dir called not just `swift` ?

mustiikhalil replied:
will be removing this gitignore

aardappel commented on tests/docker/languages/Dockerfile.testing.swift_5:
```diff
@@ -0,0 +1,8 @@
+FROM swift:5.1
```
nice!

mustiikhalil commented on Package.swift:
```diff
@@ -0,0 +1,22 @@
+// swift-tools-version:5.1
```
If this isn't in the root, we won't be able to use SPM to include the swift version in projects, unless they clone the entire repository and then try to local the swift code within their own documents.

mustiikhalil commented on Sources/FlatBuffers/FlatBuffer.swift:
```diff
@@ -0,0 +1,222 @@
+    ///   - index: index of the string in the buffer
+    ///   - count: length of the string
+    ///   - type: Encoding of the string
+    public func readString(at index: Int32, count: Int32, type: String.Encoding = .utf8) -> String? {
```
@aardappel the `FlatBuffer` is actually `ByteBuffer` and since all the languages have their Read functions in there, I added it there too. I think if it made you confused we should rename it

mustiikhalil commented on Sources/FlatBuffers/FlatbuffersErrors.swift:
```diff
@@ -0,0 +1,60 @@
+
+extension FlatbufferError: LocalizedError {
+    
+    public var errorDescription: String? {
```
we use `fatalError()` which will just quit the app or running, however we can switch to assertions if that's what the norm is

mustiikhalil commented on Sources/FlatBuffers/Struct.swift:
```diff
@@ -0,0 +1,19 @@
+    private var _bb: FlatBuffer
+    private var _postion: Int32
+    
+    public var bb: FlatBuffer { return _bb }
```
Will change this, and keep the variables public

nibres replied:
public private(set) var bb: FlatBuffer

mustiikhalil commented on include/flatbuffers/idl.h:
```diff
@@ -575,6 +575,7 @@ struct IDLOptions {
     kLobster = 1 << 13,
     kRust = 1 << 14,
     kKotlin = 1 << 15,
+    kSwift = 1 << 1,
```
that's my bad I forgot to change it at the end

mustiikhalil commented on samples/sample_binary.swift:
```diff
@@ -0,0 +1,68 @@
+    assert(monster.name == "Orc")
+    assert(monster.color == MyGame1.Sample.Color.red)
+    var vec = monster.pos
+    assert(vec?.x == 1.0)
```
okay, but that won't assert the values of `Vec` which I guess won't be a problem if we just asset the entire value

nibres commented on Sources/FlatBuffers/FlatBuffersBuilder.swift:
```diff
@@ -0,0 +1,496 @@
+
+// MARK: - Create Tables
+
+extension FlatBuffersBuilder {
```
For readability reason and grouping. Performance and behavior would be the same either putting in in class or class extension.

mustiikhalil commented on .gitignore:
```diff
@@ -115,4 +121,4 @@ dart/build/
 Cargo.lock
 .corpus**
-.seed**
+.seed**
```
Yes, I removed a newline

mustiikhalil commented:
@aardappel and @mzaks I fixed all the changes you requested yesterday:
- moved the cpp code functions inline whenever we use them once
- used auto whenever we can
- moved the swift code into dir `/swift`
- changed `FlatBuffer` to `ByteBuffer` to keep the naming consistent
- made the code use only 100 columns
- removed the errors enum since we moved everything assertions
- removed the table, struct public getters and used what @nibres suggested `public private(set) var bb: ByteBuffer`
- I explained why we should keep using the Int32() casting for some functions, and also migrated the rest to use the swift Int.

aardappel commented:
Re: errors: yes, just quitting the app on error is definitely not good behavior for a library, since that doesn't make it self-contained. We want assertions for things where it is definitely the programmers error (not a data/user error), see where the C++ code uses assertions. For where the programmer asks for a different data type than the data has stored, we want defaults, such as Null or empty string/vector/map, see what the C++ code does.

Docs should be added as part of this PR, yes.

mustiikhalil commented:
@aardappel if that's the case it should be all fixed. 
I added the documentation to this part of the PR, and moved all the required `fatalErrors()` into `assertions` and all the other issues you pointed out. Regarding the defaults in the return values, that was implemented since the start, i think we missunderstood each other regarding the casting of `Int32` it's only used when we are writing into the buffer, and it is used rarely now after you suggested changing it.
I think all the changes, you requested are done, and i also added to documentation for swift. if you have any other changes we are listening :D but other than that i think the swift implementation is ready

aardappel commented:
Ok, this looks good enough to merge to me! I haven't seen a comment from @mzaks here in a while though, I'd like his approval before we do so.

mustiikhalil commented:
@aardappel sounds good to me too! when @mzaks gives the okay, i will rebase the entire branch, and we would be good to go!

aardappel commented:
Doesn't look like it needs a rebase..

mustiikhalil commented:
I mean it wont be nice to push 64 commits into master, i think one commit will do the job

aardappel commented:
We use "squash and merge" on github, but yes, doing your own squash is preferable, as you'll get a nicer commit message :)

mzaks commented:
Sorry for the radio silence. I looked at the generated code, I see two small issues:

1. The visibility of the generated structs and static methods is not defined so it defaults to the module. I think it makes sense to make public things public, because people some times put the generated files into separate modules, or the application might be highly modularised. That sad there are also cases where people don't want generated files to leave a certain boundary of a module, so a possibility to configure in schema, or as compiler argument would be best, but IMHO the default should be `public` as this way we do less harm.
1.1 `init(_ t: Table) { _accessor = t }` is now also module protected, I think it makes sense to keep them private as I see no big advantage of exposing this API to developers.
2. I tried the code generator a couple a days ago and there a vector of unions was working for Swift. Now I get `27: error: Vectors of unions are not yet supported in all the specified programming languages.` it's ok if it is intentional though.

mustiikhalil commented:
@mzaks no worries! 
1- I will fix that, by making all of the objects generated code public
2- do you have an example? of a generated vector of unions?

mustiikhalil commented:
@aardappel I don't know why but the test case for appveyor is now failing, and travisCI is not working anymore

mzaks commented:
> 2- do you have an example? of a generated vector of unions?

The abs I am using looks like this:
```
enum Mood: byte {
	good, bad, its_something
}

table Telefon {
	number: string;
}

table URL {
	value: string;
}

table Email {
	value: string;
}

union Address {
	Telefon, URL, Email
}

table Person4 {
	valid: bool;
	name: string;
	confirmed: bool;
	moods: [Mood];
	mainAddress: Address;
	otherAddresses: [Address];
}

root_type Person4;
```

I don't have the generated code any more, but I think the API was something like this:
```
public func OtherAddressCount: Int32
public func OtherAddressType(at index: Int32) -> Address?
public func OtherAddress<T: FlatBufferObject>(at index: Int32) -> T?
```

mustiikhalil commented:
@mzaks I fixed the issue you pointed out and added test cases for it. you can find it in the following class `FlatBuffersUnionTests.testUnionVector()`

aardappel commented:
If you're going to support the newer union functionality, please make sure that you do it with the schema in `tests/union_vector`, and with similar tests to the languages there-in, instead of making yet more Swift-specific tests or schemas. The more Swift testing is set up like the other languages, the easier it will be to keep up to date in the future.

mustiikhalil commented:
Will be refactoring the test cases for the union right now

mustiikhalil commented:
@aardappel Done, we used the `union_vector.fbs` to generate the file, and then used the java test case to see if it's working

aardappel commented:
LGTM. @mzaks this ready to land?

mustiikhalil commented:
Just had to call `clang-format` on the swift cpp file

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,721 @@
+
+  void GenerateStructBody(const StructDef &struct_def, std::string *code_ptr,
+                          std::string nameprefix, int offset = 0) {
+    std::string &code = *code_ptr;
```
`code` used only once at line 133. is it really needed?

mustiikhalil replied:
well, I kinda stole the idea from csharp since the code is generated recursively. I did try removing it however that didn't work out as expected. if you have better ideas I am all ears.

mustiikhalil replied:
just realized we can directly use `code_`

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,721 @@
+      auto name = nameprefix + Name(field);
+      auto type = GenTypeBasic(field.value.type, false);
+      const auto &field_type = field.value.type;
+      if (field.deprecated) continue;
```
`deprecated` check can be moved up, just after `auto &field`.

mustiikhalil replied:
Sure!

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,721 @@
+    std::string spacing = "\t\t";
+    if (!struct_def.fixed && struct_def.has_key) {
+      code_.SetValue("VALUENAME", struct_def.name);
+      code_.SetValue("VOFFSET", NumToString(key_field->value.offset));
```
is key_field not-null here? what will happen if all fields are deprecated?

mustiikhalil replied:
so you think a check would be nice before entering the if statement? 
`if (key_field != nullptr && !struct_def.fixed && struct_def.has_key) {`

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,721 @@
+          "{{VOFFSET}}, fbb: fbb.buffer), fbb: fbb.buffer) < 0 } ";
+      code_ += spacing + "return fbb.createVector(ofOffsets: off)";
+      code_ += "\t}";
+      GenLookup(*key_field);
```
is key_field not-null here? what will happen if all fields are deprecated?

mustiikhalil replied:
this would be the same guard at the top

vglavnyy commented:
I can clone and check idl_gen_swift.cpp with static tools before landing.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,721 @@
+  }
+
+  void LowerCase(std::string *value) {
+    transform(value->begin(), value->end(), value->begin(), ::tolower);
```
MSVC2019 compilation failed with error:
>include\algorithm(1310,24): error C2220: the following warning is treated as an error
>warning C4244: '=': conversion from 'int' to 'char', possible loss of data

You can see ToUpper implementation in idl_gen_cpp or [here](https://en.cppreference.com/w/cpp/string/byte/tolower).
>Like all other functions from <cctype>, the behavior of std::tolower is undefined if the argument's value is neither representable as unsigned char nor equal to EOF. To use these functions safely with plain chars (or signed chars), the argument should first be converted to unsigned char

mustiikhalil commented:
@vglavnyy that would be nice! thank you! tell me what do you think of the review

mustiikhalil commented:
@vglavnyy fixed all the issues you mentioned

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+
+inline void LowerCase(std::string *value) {
+  transform(value->begin(), value->end(), value->begin(),
+            [](unsigned char c) { return std::tolower(c); });
```
[](char c) { return static_cast<char>(::tolower(static_cast<unsigned char>(c))); });

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+                 const std::string &file_name)
+      : BaseGenerator(parser, path, file_name, "", "."),
+        cur_name_space_(nullptr) {
+    static const char *const keywords[] = {
```
The private `should_generate_extension` not inited in ctor.
Why you need this pointer? See my comment on line 704

vglavnyy replied:
V730 [CWE-457] Not all members of a class are initialized inside the constructor. Consider inspecting: should_generate_extension, namespace_depth. idl_gen_swift.cpp 32

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+      code_ += "extension {{EXTENSION}} {";
+    }
+    if (new_size != common_prefix_size) { code_ += ""; }
+    should_generate_extension = &is_extension;
```
`is_extension` is **local** variable declared at 687.
I would recommend removing the pointer `should_generate_extension` from the code.
I don't understand why this variable is pointer, not value.

vglavnyy replied:
PVS found it too:
V506 [CWE-562] Pointer to local variable 'is_extension' is stored outside the scope of this variable. Such a pointer will become invalid. idl_gen_swift.cpp 704

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    return SaveFile(filename.c_str(), final_code, false);
+  }
+
+  void mark(const std::string str) {
```
V801 Decreased performance. It is better to redefine the first function argument as a reference. Consider replacing 'const .. str' with 'const .. &str'. idl_gen_swift.cpp 94

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    code_ += "}\n";
+  }
+
+  void GenerateStructBody(const StructDef &struct_def, std::string nameprefix,
```
V813 Decreased performance. The 'nameprefix' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 119

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    }
+  }
+
+  void GenerateStructArgs(const StructDef &struct_def, std::string *code_ptr,
```
V813 Decreased performance. The 'nameprefix' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 142

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+
+namespace swift {
+
+inline std::string GenIndirect(std::string reading) {
```
V813 Decreased performance. The 'reading' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 12

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+  return "{{ACCESS}}.indirect(" + reading + ")";
+}
+
+inline std::string GenArrayMainBody(std::string optional = "") {
```
V813 Decreased performance. The 'optional' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 16

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+
+  std::string GenOffset() { return "let o = {{ACCESS}}.offset({{OFFSET}}); "; }
+
+  std::string GenReaderMainBody(std::string optional = "") {
```
V813 Decreased performance. The 'optional' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 560

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    return "\tpublic var {{VALUENAME}}: {{VALUETYPE}}" + optional + " { ";
+  }
+
+  std::string GenReader(std::string type, std::string at = "{{OFFSET}}") {
```
V813 Decreased performance. The 'type', 'at' arguments should probably be rendered as constant references. idl_gen_swift.cpp 564

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    return "{{ACCESS}}.readBuffer(of: {{" + type + "}}.self, at: " + at + ")";
+  }
+
+  std::string GenConstructor(
```
V813 Decreased performance. The 'offset' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 569

Usually, the default initializer is short. This function called without arguments only once.
Please don't use default arguments in a declaration if a function called in short form less than 2 times.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    return "{{VALUETYPE}}({{ACCESS}}.bb, o: " + offset + ") }";
+  }
+
+  std::string GenMutate(std::string offset, std::string get_offset = "",
```
V813 Decreased performance. The 'offset', 'get_offset' arguments should probably be rendered as constant references. idl_gen_swift.cpp 573

This function called with the default `get_offset` only once.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+    return "{{VALUETYPE}}(rawValue: {{CONSTANT}})!";
+  }
+
+  std::string GenEnumConstructor(std::string at = "{{OFFSET}}") {
```
V813 Decreased performance. The 'at' argument should probably be rendered as a constant reference. idl_gen_swift.cpp 592

It called with the default only once.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,719 @@
+                               Name(*type.enum_def));
+      if (type.base_type == BASE_TYPE_BOOL) return "Bool";
+    }
+    return swift_type[type.base_type];
```
V2006 Implicit type conversion from enum type to integer type. idl_gen_swift.cpp 633

Better to use explicit static_cast here.

mustiikhalil commented:
@vglavnyy I think all the issues are resolved regarding cpp! 
@mzaks i changed the reader function to loading the scalar value from the buffer directly instead of using a pointer to read it

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -9,30 +9,34 @@ namespace flatbuffers {
 namespace swift {
 
-inline std::string GenIndirect(std::string reading) {
+inline std::string GenIndirect(const std::string reading) {
```
It should be const reference: `const std::string &reading`.
The same for others `const std::string`, they should be declared as reference `const std::string&`.

`const std::string` is a pass-by-value semantic, if inlining failed then a copy of string will be created.
With reference, a compiler will pass the address of a string.

mustiikhalil replied:
@vglavnyy so if we inline a function in cpp, we should use references instead of just passing a std::string.
I’m still kinda learning cpp so forgive me for that

vglavnyy replied:
It doesn't depend on inlining. Inline word is a recommendation only which can be ignored by a compiler.
```cpp
void foo(const std::string s) {
 // s is immutable string
}

void bar(const std::string s) {
 // s is reference to immutable string
}

void test() {
 std::string text = "bar and foo";
 foo(text); // foo(std::string(text)) - text will be cloned and passed into foo
 bar(text); // bar(&text) - pass address of text without clone
}
```

mustiikhalil replied:
Thanks for the explanation!

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,715 @@
+    return EscapeKeyword(MakeCamel(def.name, false));
+  }
+
+  const Namespace *cur_name_space_;
```
Please, move `cur_name_space_` into the declaration area of the class.

mustiikhalil commented:
@vglavnyy I refactored the code according to the changes you asked for, and also besides moving the `GenTableReader()` into a another function I also moved the generation of vectors into another function which hopefully would help readability even more.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -210,46 +219,17 @@ class SwiftGenerator : public BaseGenerator {
-                : "offset: {{VALUENAME}}, at: {{OFFSET}})  }";
-        code_ += offset_type + builder_string + reader_type;
-      }
+      GenTableWriterFields(field, it - struct_def.fields.vec.begin());
```
> warning C4244: 'argument': conversion from '__int64' to 'const int', possible loss of data

1. static_cast<int>()
2. add `position` variable: GenTableWriterFields(field, position++);

vglavnyy commented:
Thanks, looks nice.

mustiikhalil commented:
@vglavnyy Glad! and thanks for the review! i did learn alot!

mustiikhalil commented:
1- I had to fix an issue where swift doesn't understand the constant that we are passing to it for the Enums. 
2- Moved adding a scalar into the buffer to use memcpy instead of the swift implementation

mustiikhalil commented:
@mzaks what do you think of the PR now? I've fixed everything on swift, and the code gen.

vglavnyy commented on src/idl_gen_swift.cpp:
```diff
@@ -0,0 +1,715 @@
+      const auto &field_type = field.value.type;
+      if (IsStruct(field.value.type)) {
+        GenerateStructBody(*field_type.struct_def,
+                           (nameprefix + field.name.c_str()),
```
@mustiikhalil It is not fixed.
1. `nameprefix` is `std::string`.
2. `field.name`  is `std::string`.

`field.name.c_str()` -> `const char*` - for what?

mustiikhalil replied:
@vglavnyy my bad. I think I reset the changes that fixed that.

aardappel commented:
@vglavnyy thanks for the review! If only we could run PVS-studio as CI :)

vglavnyy commented:
I will try to add PVS and Coverity Scan to Travis-CI :)

mustiikhalil commented:
just FYI, I did try using the following `pushd "$(dirname $0)" >/dev/null` but it didn't work and it always resulted in the following error `SwiftTest.sh: 4: SwiftTest.sh: pushd: not found` so I made it use `swift_dir="pwd"` directly

mustiikhalil commented:
@aardappel and @mzaks I think this PR is finalized now, whenever @mzaks has time to review the swift code that would be amazing. I've already tested it on linux, macOS and iOS. I've fixed all the memory issues that we were facing. I've used Xcode's the following to debug the entire project:
1- Malloc Stack and fixed all the allocation issues that we faced
2- Address Sanitizer that makes sure that we aren't writing or reading from memory that doesn't belong the buffers. 
all of them passed.

mzaks commented:
Generally thing already looks quite great. I have another few things, which are not blockers, but seem strange to me and could be addressed now or maybe in sub sequential releases:


1. In `union_vector_generated.swift` we have following API `public func mutate(mainCharacterType: Character)` which implies that we can mutate the type of the character inline. IMHO this is problematic as the actual value can't be mutated and the layout of tables included in the union can be very different. e.g. `Character.Mulan` is of type `Attacker` and `Character.Other` is of type `string` meaning changing just the type of a union can do more harm than good in my opinion.
2. Same file, same class, we have `public var charactersTypeCount: Int32` and `public var charactersCount: Int32` which is not bad it is just redundant as both should return the same value.
3. Same file, same class, `public static func add(charactersType: Offset<UOffset>, _ fbb: FlatBufferBuilder)` just a small thing, I think it make sense to name the first parameter `characterTypes` as other wise it feels like it should be same type for multiple characters, where it is actually different types per character. Maybe a more descriptive name would be even better, as we the type does not directly communicate that we expect an offset to a vector.
4. In `moster_test_generated.swift` I see that we support the `(key)` annotation, which is awesome, but the API `public static func lookupByKey(vector: Int32, key: UInt64, fbb: ByteBuffer) -> Referrable?` and other `lookupByKey` static methods should probably be of visibility `fileprivate` or at least `internal ` as I don't think end users will be able to do something useful with this API.

All in all great job @mustiikhalil!

mzaks commented:
One more thing which could make Swift support better and probably could benefit also other languages as well. Swift package manager works best if the git repos root contains the `Package.swift` file. In our case the `Package.swift` feel is in the `swift` folder, so we can't point SPM directly to the GitHub repo. There is a solution by using git subtree https://www.atlassian.com/git/tutorials/git-subtree. I stumbled upon this solution just recently myself by reading following blog post (https://neogeek.dev/creating-custom-packages-for-unity-2018.3/).

mustiikhalil commented:
@mzaks so regarding the `package.swift`, we did have this conversation with @aardappel and he requested it to be within the directory `swift` since the other option was a git submodule. However if the git subtree would be helpful we should look into it.
1- You are absolutely right! Completely forgot about that one, I will be removing all the unions from the mutations.
2- All the languages actually produce it all the time, so I think we should also keep it that way too.
3- 
3.1 the actual fbs file has the name of the element as `characters` and when we generate the field in other languages it's always handled like this `CharactersType` which is similar to what we have in swift. 
3.2: what do you think of the following then `addVectorOf(charactersType:)`? 
4- that's a great idea, we should make it `fileprivate`since we don't want the user to really use it

aardappel commented:
@mzaks would prefer to keep package file in subdir if at all possible.

mustiikhalil commented:
@aardappel, @mzaks fixed all the issues you mentioned yesterday:
1- Removed the union mutation.
2- Renamed the vector functions to `addVectorOf(elements:)`, however the naming of the elements is the same as all the other languages.
3- `fileprivate` for the lookup functions.

However, I didn't change the count for the unions arrays since all the languages do the same thing, so I kept that to adhere to the other APIs. and didn't touch the `package.swift`

mzaks commented:
@aardappel as far as I can tell in order to make subtree work, we need something like following command:
```
git subtree push --prefix swift origin swift
```
To create a new branch `swift`, which contains only the code from `swift` folder. This branch can be used for releases of Swift Library and also as a URL for Swift Package Manager. We would need to enhance the build and release pipeline though. But then we could use the same technique for Rust, Dart, etc... And move `package.json` and `pom.xml` into subfolders.

mzaks commented:
@mustiikhalil one more small thing, you probably what to `cd` back here:
https://github.com/mustiikhalil/flatbuffers/blob/swift-implementation/tests/TestAll.sh#L64
Now it is not a problem as swift is the last test but it will be bad for people who want to add another case.

mustiikhalil commented:
@mzaks that's done :D so hopefully the PR looks clean now! and the swift implementation would be good to go! 
I did try that `git subtree` and that would actually be great! since it would enable all the languages to be in the same repository in different dir, however act as if they are maintained in different repositories! [see example](https://github.com/mustiikhalil/learn-rust) I made also a small example @aardappel so you can actually see it's going to be worth it. the repo has all directories, however the branch `ownership` only has the `ownership` files only. however, I am not 100% sure how maintaining a branch for each language would be.

aardappel commented:
Yup, would totally be a fan of moving other language cruft into subdirs as well.. but that's probably for another PR.

So.. ready for me to merge?

mustiikhalil commented:
i think so, unless @mzaks would say otherwise

aardappel commented:
Ok, time to merge then!
Anything else can be in a follow-up.
Thanks @mustiikhalil for your hard work, and @mzaks and @vglavnyy  for reviewing :)
